### PR TITLE
Client Database Transactions and Remove DbBatch

### PR DESCRIPTION
**Pull request by m1sterc001guy**
This PR refactors the client to use the new database transaction trait. The client does not rely on snapshots or rolling back of transactions.

This PR also removes DbBatch and the only API for interacting with the database.

### CODE REVIEW

File: client/client-lib/src/lib.rs [hunk 1]
```diff
@@ -871,32 +854,33 @@ impl Client<GatewayClientConfig> {
...
-                &contract,
-            )
-            .expect("DB error");
+        let mut dbtx = self.context.db.begin_transaction();
+        dbtx.insert_entry(
+            &OutgoingContractAccountKey(contract.contract.contract_id()),
+            &contract,
+        )
+        .expect("DB error");
+        dbtx.commit_tx().expect("DB Error");
```
Comment (elsirion) [inline client/client-lib/src/lib.rs:863]: The client might do things in parallel, is it actually true that there can be no DB errors? `expect("DB error")` was used in places where it essentially means "our DB isn't reachable", which for an embedded DB essentially means your hard drive went missing/is on fire imo. If an error can actually reasonably occur we should handle it instead.

File: modules/fedimint-ln/src/lib.rs [hunk 2]
```diff
@@ -275,6 +276,7 @@ impl FederationModule for LightningModule {
         let account_db_key = ContractKey(input.contract_id);
         let mut contract_account = self
             .db
+            .begin_transaction()
```
Comment (elsirion) [inline modules/fedimint-ln/src/lib.rs:279]: I know this replicates old behavior, but shouldn't we just switch to only reading from the DB tx and actually remove the DB reference from the module all together?

**Review by elsirion — CHANGES_REQUESTED** [review]
Goes into the right direction, I'm wondering if you see it as a good atomic step or if we should properly migrate the modules over. I fear it's easier to miss a place where we should be reading from the main tx but create a new one instead after this change.

File: client/client-lib/src/lib.rs (hunk 1 above)
Comment (m1sterc001guy) [reply client/client-lib/src/lib.rs:863]: Yes, you're right, I think it is possible for conflicts to happen here. I'll revisit these and try to handle them properly.

File: modules/fedimint-ln/src/lib.rs (hunk 2 above)
Comment (m1sterc001guy) [reply modules/fedimint-ln/src/lib.rs:279]: The PR actually removes all other functions from the trait besides begin_transaction(). So the only thing to do with a database reference is to create a new transaction.

The rocksdb implementation needs a database reference to create the transaction. I'm open to suggestions if you have ideas on encapsulating this (perhaps creating a wrapper like this: self.create_transaction()....).

**Comment by m1sterc001guy** [comment]
What do you mean by properly migrate the modules over? I could reduce the scope of the PR if you want, perhaps limiting it to only client changes (but keep DbBatch and the current database trait). I don't think the current change introduces any new semantics, since the old database trait would also be reading from a snapshot. I think this PR just makes it explicit.

**Review by elsirion — APPROVED** [review]

**Comment by elsirion** [comment]
My initial concerns were unfounded as explained above, let's iterate on master.

**Pull request merged**